### ISSUE TITLE
Flow API accumulates points when tab is not focussed

### ISSUE THREAD

**Issue by yourivdlans** [masayuki0812/c3#1097]
I've created a chart which shows live data in a chart using flow. However when the tab with the chart in it is not focussed the chart will accumulate data points until you return to the tab. Sometimes slowing down your browser until the chart has been rendered.

You can simulate this by opening the fiddle: http://jsfiddle.net/2camvme7/
And letting it run while you open up a new tab and stay there for a few minutes. When you return to the fiddle you'll see that the chart now has a lot more points. But when you keep the tab with the fiddle open it will work as its supposed to.

Only tested with Chrome (41.0.2272.104).

**Comment by ianjennings** [comment]
Having the same issue here. If enough points accumulate it will crash your browser.

**Comment by mrowa** [comment]
I've created a simple override, can't post the code though - stop flows when the tab is not active (I use VisibilityChange API) and if you need only some amount of points, a constant-length queue (or queue filtering older-than data) takes the data and discards already irrelevant ones. With some queue + data length math you can override additions to chart only with data still in the queue.

**Comment by yourivdlans** [comment]
Also tried implementing that approach but it would still crash sometimes.

Is there a way to extract your piece of code so we can have a look? Would be really helpful!

**Comment by mrowa** [comment]
Sorry, I won't have time to do so in next few weeks. But we've dropped C3 and used Rickshaw (which is inactive) for real-time charts instead. 

It's (too) based on D3, but it's optimized no to do too many CPU-intensive operations, like fluid reflows. Our application came to a crawl with just 4 medium sized C3 charts, hence the change. 

We've checked a few of svg charting library, e.g. Highcharts has similar way of work as C3... so we used Rickshaw. If those would become too slow in long term, we'll switch to something canvas-based, or try our luck with bare D3.

Finally, about the solution: 
  1. There's an issue in C3 for IE10/11 that makes C3 hog memory and crash, if tab was switched, when in animation (https://github.com/masayuki0812/c3/issues/1183), this solution is not an override.
  2. C3 may be quite slow with flows (especially if there's a few of them on one screen), don't flow new data if you are sure the old flow is not finished.
  3. Do not run flows when the page is inactive, you can use PageVisibility API to check for being active.
  4. Create max-size/max-length FIFO queue (or use lib) which would be your data source for charts (I used row data instead of columns), rows go in, old rows go out.
  5. When the page is inactive, stop flows, gather readings in queue and remove unnecessary old ones. After first activity, add all points in queue to the chart.

This is how I've done it, it worked ok.

**Comment by adamlwgriffiths** [comment]
This is still an issue.

Seems it's related to this code here:
https://github.com/masayuki0812/c3/blob/79c43c27bd70b532c5864995804c02ddb1a04043/src/core.js#L650
https://github.com/masayuki0812/c3/blob/98769492d07b6103bfc30a0254ccb1e1ec1cca50/c3.js#L655

C3 is doing this to avoid transitions wasting CPU time.
The proper way to handle it would be to update the graph without transitions while the tab is hidden.

**Comment by petersandor** [comment]
I'm experiencing the same issue @0.4.11.

**Comment by MartinWetterwald** [comment]
I'm also experiencing the issue with 0.4.11.
Sometimes I forget and let the tab idle during 24h. A lot of points accumulate. When I come back the next day my Firefox is so stuck that I need to SIGKILL it.

**Comment by ianjennings** [comment]
I've implemented a workaround based on @mrowa's suggestion using the visibility.js. 
https://github.com/ai/visibilityjs

When the page is out of focus, I destroy the graph and accumulate point into a queue. When the focus returns, I rebuild the graph using accumulated points.
https://github.com/pubnub/eon-chart/blob/master/pubnub-c3.js#L251

It should be possible to implement this same approach within C3.js

**Comment by flavioespinoza** [comment]
I used a similar solution to @ianjennings;
I used this to get tab visibility http://stackoverflow.com/a/25628619/3310801

Then in my JS file:

`if (document[visibilityState] === "visible") {
                // visible
                __chart.flow({
                  columns: [
                    ['x', '8:00:07 PM'],
                    ['cpu usage', 0.24],
                    ['mem usage', 0.50]
                  ],
                  duration: 1000
                });
              } else {
                // hidden
                __chart.load({
                  columns: [
                    ['x', '8:00:01 PM', '8:00:02 PM', '8:00:03 PM', '8:00:04 PM', '8:00:05 PM', '8:00:06 PM', '8:00:07 PM', ],
                    ['cpu usage', 0.24, 0.34, 0.14, 0.24, 0.34, 0.14, 0.24],
                    ['mem usage', 0.54, 0.34, 0.34, 0.14, 0.64, 0.24, 0.50]
                  ]
                });

              }`

**Comment by flavioespinoza** [comment]
I used a similar solution to @ianjennings: 
http://codepen.io/flavioespinoza/pen/XMdaBR?editors=0010

**Comment by yunyu** [comment]
I adapted a solution off of @adamlwgriffiths, which occasionally failed with "Cannot read property 'x' of null" if there were more than 5 items in the buffer. I ended up collapsing the items in the buffer to mitigate it, which should be good enough for a general workaround (if your data has <5 flow unique length values): https://github.com/yunyu/vertx-metrics-frontend/blob/f53f565e6b42a1ffae3fcbcdeb08aa6cdf3e17fd/src/components/c3/FlowedLine.vue#L25

**Comment by adamlwgriffiths** [comment]
Yeah, even with my work around I would still get serious performance issues.
I've since dropped using C3 because of this.
